internal/alloycli/cluster_builder.go

```diff
@@ -68,24 +70,39 @@ func buildClusterService(opts clusterOptions) (*cluster.Service, error) {
 		return nil, err
 	}
 
-	switch {
-	case len(opts.JoinPeers) > 0 && opts.DiscoverPeers != "":
-		return nil, fmt.Errorf("at most one of join peers and discover peers may be set")
-
-	case len(opts.JoinPeers) > 0:
-		config.DiscoverPeers = newStaticDiscovery(opts.JoinPeers, listenPort, opts.Log)
-
-	case opts.DiscoverPeers != "":
-		discoverFunc, err := newDynamicDiscovery(config.Log, opts.DiscoverPeers, listenPort)
+	// New, refactored and improved peer discovery.
+	// TODO(alloy/#1274): Remove the old peer discovery code once this becomes GA.
+	if opts.EnableDiscoveryV2 {
+		config.DiscoverPeers, err = discovery.NewPeerDiscoveryFn(discovery.Options{
+			JoinPeers:     opts.JoinPeers,
+			DiscoverPeers: opts.DiscoverPeers,
+			DefaultPort:   listenPort,
+			Logger:        opts.Log,
+			Tracer:        opts.Tracer,
+		})
 		if err != nil {
 			return nil, err
 		}
-		config.DiscoverPeers = discoverFunc
+	} else {
+		switch {
+		case len(opts.JoinPeers) > 0 && opts.DiscoverPeers != "":
+			return nil, fmt.Errorf("at most one of join peers and discover peers may be set")
+
+		case len(opts.JoinPeers) > 0:
+			config.DiscoverPeers = newStaticDiscovery(opts.JoinPeers, listenPort, opts.Log)
+
+		case opts.DiscoverPeers != "":
+			discoverFunc, err := newDynamicDiscovery(config.Log, opts.DiscoverPeers, listenPort)
+			if err != nil {
+				return nil, err
+			}
+			config.DiscoverPeers = discoverFunc
 
-	default:
-		// Here, both JoinPeers and DiscoverPeers are empty. This is desirable when
-		// starting a seed node that other nodes connect to, so we don't require
-		// one of the fields to be set.
+		default:
+			// Here, both JoinPeers and DiscoverPeers are empty. This is desirable when
+			// starting a seed node that other nodes connect to, so we don't require
+			// one of the fields to be set.
+		}
```

Comment:
The diff doesn't make it easy to see that this `else` is basically the original cluster peer discovery code.